nemo_aligner/models/nlp/gpt/gpt_steerlm_model.py

```diff
@@ -0,0 +1,357 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from importlib.metadata import version
+from itertools import chain
+from typing import List, Optional, Tuple, Union
+
+import hydra
+import torch
+from apex.transformer.pipeline_parallel.utils import get_micro_batch_size, get_num_microbatches
+from megatron.core import parallel_state
+from megatron.core.pipeline_parallel.schedules import get_forward_backward_func
+from omegaconf.dictconfig import DictConfig
+from pkg_resources import packaging
+from pytorch_lightning.trainer.trainer import Trainer
+from torch.distributed import barrier
+
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
+from nemo.collections.nlp.modules.common.megatron.utils import (
+    average_losses_across_data_parallel_group,
+    get_iterator_k_split,
+)
+from nemo.collections.nlp.modules.common.text_generation_strategy import TextGenerationStrategy
+from nemo.collections.nlp.modules.common.text_generation_utils import (
+    get_default_length_params,
+    get_default_sampling_params,
+)
+from nemo.collections.nlp.modules.common.transformer.text_generation import LengthParam, OutputType, SamplingParam
+from nemo.collections.nlp.parts.mixins.nlp_adapter_mixins import NLPAdapterModelMixin
+from nemo.collections.nlp.parts.utils_funcs import get_last_rank
+from nemo.utils import logging
+from nemo_aligner.models.alignable_interface import SupervisedInterface
+from nemo_aligner.models.nlp.gpt.gpt_sft_model import GPTSFTModel
+from nemo_aligner.utils.train_utils import (
+    finish_validation_step,
+    grad_reductions,
+    prepare_for_training_step,
+    prepare_for_validation_step,
+    set_eval,
+    set_sync_funcs,
+    set_train,
+)
+from nemo_aligner.utils.utils import configure_batch_sizes
+
+
+def get_batch(data_iterator, tuning):
+    """Generate a batch."""
+
+    # Broadcast data.
+    if data_iterator is not None:
+        data = next(data_iterator)
+    else:
+        data = None
+
+    # return batch for GPT SFT
+    if tuning:
+        return data
+
+    batch = {
+        "tokens": data["tokens"],
+        "labels": data["labels"],
+        "loss_mask": data["loss_mask"],
+        "attention_mask": data["attention_mask"],
+        "position_ids": data["position_ids"],
+    }
+    if "weight" in data:
+        batch.update({"weight": data["weight"]})
+    if "avg_num_valid_tokens_in_ub" in data:
+        batch.update({"avg_num_valid_tokens_in_ub": data["avg_num_valid_tokens_in_ub"]})
+    return batch
+
+
+class GPTSteerLMModel(GPTSFTModel):
+    def loss_func(self, loss_mask, output_tensor):
+        # loss is the nll loss for each token in the batch
+        losses = output_tensor.float()
```

Comment:
nit: https://github.com/NVIDIA/Megatron-LM/blob/54b4de5b0409313ad11b549f488cb5dc5193ef03/megatron/core/transformer/module.py#L175

already converts it to fp32